### PR TITLE
Change Twig_Loader_Chain to accept Traversable

### PR DESCRIPTION
**Pull request by enumag**

### CODE REVIEW

File: lib/Twig/Loader/Chain.php [hunk 1]
```diff
@@ -22,8 +22,12 @@
     /**
      * @param Twig_LoaderInterface[] $loaders
      */
-    public function __construct(array $loaders = array())
+    public function __construct($loaders = array())
     {
+        if (!is_array($loaders) && !$loaders instanceof \Traversable) {
```
Comment (linaori) [inline lib/Twig/Loader/Chain.php:27]: What about `iterable`?

Comment (nicolas-grekas) [reply lib/Twig/Loader/Chain.php:27]: PHP 7.1 only

**Comment by enumag** [comment]
Can this be merged? The tests failure is not related to my change.

**Comment by nicolas-grekas** [comment]
addLoader will fail when a non-array is injected I suppose, shouldn't it throw an exception in this case?

**Comment by enumag** [comment]
Huh? It works fine with non-array iterable... You mean iterable with different items than Twig_LoaderInterface? In my opinion it's not necessary to add such exception. Useless balast code if you ask me.

**Comment by nicolas-grekas** [comment]
`$this->loaders[] = ...` <- this doens't work for all `Traversable`, so there is no gurantee it will work at all with any Traversable, that's what I meant.

**Comment by enumag** [comment]
I have no idea what you mean. `$this->loaders` is always array even with my change.

**Comment by nicolas-grekas** [comment]
Oh, OK, got it sorry, the Traversable is turned to an array in the constructor with the foreach. I thought it would be kept as is for laziness purpose.

**Comment by stof** [comment]
Iterating in the constructor removes the benefit of using a lazy-loading iterator (especially if your first loader is able to load most of your templates.
It would be better to keep lazyness, and turn the iterator into an array only when adding a new loader later (projects wanting to benefit from lazyness would then have to take care to provide all loaders in the constructor, but this is fine IMO)

**Comment by enumag** [comment]
@stof Possibly but that's not really the aim of this PR.

**Comment by stof** [comment]
Well, you are accepting Traversable, but without any benefit for passing a Traversable rather than an array.

**Comment by enumag** [comment]
The ability to pass Traversable is the benefit for me. But alright, I'll see what I can do about the laziness.

**Comment by fabpot** [comment]
@enumag Have you made any progress on the laziness? Are you still interested in finishing this pull request?

**Comment by enumag** [comment]
@fabpot I solved it differently so not interested anymore.

**Pull request closed without merging**